### PR TITLE
FIX KUBE INSTALLATION

### PR DESCRIPTION
**Pull request by amiraaz**
+ FIX DEPLOYEMENT AND LDAP FILES
+ ADD DEVELOPPEMENT GUIDE
+ ADD CONTRIBUTING GUIDE

### CODE REVIEW

File: Command.md
Comment (evrardjp) [inline Command.md:1]: Please merge command.md into contributing.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+## Table of Contents
+
+- [I Have a Question](#i-have-a-question)
+- [I Want To Contribute](#i-want-to-contribute)
+- [Reporting Bugs](#reporting-bugs)
+- [Suggesting Enhancements](#suggesting-enhancements)
+
+
+## I Have a Question
+
```
Comment (evrardjp) [inline CONTRIBUTING.md:24]: We probably want to stop accepting issues here, as we don't follow them. Maybe say drop an email to (give the mailing list address)

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- Make sure that you are using the latest version.
+- Determine if your bug is really a bug and not an error on your side e.g. using incompatible environment components/versions (Make sure that you have read the [documentation](https://github.com/ca-gip/kubi). If you are looking for support, you might want to check [this section](#i-have-a-question)).
+- To see if other users have experienced (and potentially already solved) the same issue you are having, check if there is not already a bug report existing for your bug or error.
+- Collect information about the bug:
+  - Stack trace (Traceback)
+  - OS, Platform and Version (Windows, Linux, macOS, x86, ARM)
+  - Version of the interpreter, compiler, SDK, runtime environment, package manager, depending on what seems relevant.
+  - Possibly your input and the output
+  - Can you reliably reproduce the issue? And can you also reproduce it with older versions?
+
```
Comment (evrardjp) [inline CONTRIBUTING.md:70]: Add:

- "What did you have as result, and what did you expect"
- "Give everything we need to reproduce the issue (a test if possible)"

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+<!-- omit in toc -->
+
+
+We use GitHub issues to track bugs and errors. If you run into an issue with the project:
```
Comment (evrardjp) [inline CONTRIBUTING.md:74]: Maybe to remove too

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+- Open an [Issue](https://github.com/ca-gip/kubi/issues/new). (Since we can't be sure at this point whether it is a bug or not, we ask you not to talk about a bug yet and not to label the issue.)
+- Explain the behavior you would expect and the actual behavior.
+- Please provide as much context as possible and describe the *reproduction steps* that someone else can follow to recreate the issue on their own. This usually includes your code. For good bug reports you should isolate the problem and create a reduced test case.
+- Provide the information you collected in the previous section.
+
```
Comment (evrardjp) [inline CONTRIBUTING.md:80]: The removal of the issue text I mentioned above stops here.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Once it's filed:
+
+- The project team will label the issue accordingly.
```
Comment (evrardjp) [inline CONTRIBUTING.md:83]: Remove L83

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+- A team member will try to reproduce the issue with your provided steps. If there are no reproduction steps or no obvious way to reproduce the issue, the team will ask you for those steps and mark the issue as `needs-repro`. Bugs with the `needs-repro` tag will not be addressed until they are reproduced.
```
Comment (evrardjp) [inline CONTRIBUTING.md:84]: L84 Keep only the first sentence and say "we will contact you back"

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+### Suggesting Enhancements
+
+This section guides you through submitting an enhancement suggestion for kubi, **including completely new features and minor improvements to existing functionality**. Following these guidelines will help maintainers and the community to understand your suggestion and find related suggestions.
+
+<!-- omit in toc -->
+#### Before Submitting an Enhancement
+
+- Make sure that you are using the latest version.
+- Read the [documentation](https://github.com/ca-gip/kubi) carefully and find out if the functionality is already covered, maybe by an individual configuration.
+- Perform a [search](https://github.com/ca-gip/kubi/issues) to see if the enhancement has already been suggested. If it has, add a comment to the existing issue instead of opening a new one.
```
Comment (evrardjp) [inline CONTRIBUTING.md:99]: Remove this

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+<!-- omit in toc -->
+#### How Do I Submit a Good Enhancement Suggestion?
+
+Enhancement suggestions are tracked as [GitHub issues](https://github.com/ca-gip/kubi/issues).
```
Comment (evrardjp) [inline CONTRIBUTING.md:104]: Remove this content, keep the title. Just add "propose a PR, and contact us for review".

File: Command.md [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+# Developer guide
+
+This guide helps you get started developing kubi
+
+Make sure you have the following dependencies installed before setting up your developer environment:
+
+ - Git
+ - Jq
+ - Go
+ - In this context we will use a kind cluster for the local deployment of Kubi : cluster kind {1.23 or 124}
```
Comment (evrardjp) [inline Command.md:11]: Version is incorrect ?

File: Command.md [hunk 10]
```diff
@@ -0,0 +1,309 @@
+<!-- omit in toc -->
+# Developer guide
+
+This guide helps you get started developing kubi
+
+Make sure you have the following dependencies installed before setting up your developer environment:
+
+ - Git
+ - Jq
+ - Go
```
Comment (evrardjp) [inline Command.md:10]: openssl, curl, and wget seems to be required, but are not listed. 

Don't we need docker to build an image too?

File: Command.md [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+   cfssljson -version
+  
+   VERSION=$(curl --silent "https://api.github.com/repos/cloudflare/cfssl/releases/latest" | grep '"tag_name"' | sed -E 's/.*"([^"]+)".*/\1/')
+   VNUMBER=${VERSION#"v"}
+   wget https://github.com/cloudflare/cfssl/releases/download/${VERSION}/cfssl_${VNUMBER}_linux_amd64 -O cfssl
+   chmod +x cfssl
+   sudo mv cfssl /usr/local/bin
+ ```
+    
+ - create certificat
```
Comment (evrardjp) [inline Command.md:47]: Missing "e" at the end of certificate

File: Command.md [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+
+ - Get the source code https://github.com/ca-gip/kubi
+  
+ - install go  https://go.dev/
+    
+ - Create kind cluster with version 1.24-1.26 https://kind.sigs.k8s.io/docs/user/quick-start/
+  
+ - kubi private key
+ ```
+ openssl ecparam -genkey -name secp521r1 -noout -out /tmp/ecdsa-key.pem
```
Comment (evrardjp) [inline Command.md:23]: /tmp/ should not be used, except if you are using mktemp.

Please adapt instructions. 

Can we make this in a single script ?

File: Command.md [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+ ```
+  
+ - kubi public key
+ ```
+ openssl ec -in /tmp/ecdsa-key.pem -pubout -out /tmp/ecdsa-public.pem
+ ```
+  
+ - install CFSSL tools
+ ```
+   VERSION=$(curl --silent "https://api.github.com/repos/cloudflare/cfssl/releases/latest" | grep '"tag_name"' | sed -E 's/.*"([^"]+)".*/\1/')
```
Comment (evrardjp) [inline Command.md:33]: Couldn't this be containerized to avoid installing binaries into the host?

File: Command.md [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+   VNUMBER=${VERSION#"v"}
+   wget https://github.com/cloudflare/cfssl/releases/download/${VERSION}/cfssljson_${VNUMBER}_linux_amd64 -O cfssljson
+   chmod +x cfssljson
+   sudo mv cfssljson /usr/local/bin
```
Comment (evrardjp) [inline Command.md:37]: Wrong place for installing optional binaries, see posix lsb

File: Command.md [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+  - Create a secret for the deployment
+  ``` 
+  kubectl -n kube-system create secret tls kubi --key server-key.pem --cert server.crt
+  kubectl -n kube-system create secret generic kubi-encryption-secret --from-file=/tmp/ecdsa-key.pem --from-file=/tmp/ecdsa-public.pem
+  kubectl -n kube-system create secret generic kubi-secret  --from-literal ldap_passwd='password'
+  ```
+
+  - Deploy manifest (CRD, prerequisites,local-config) of kubi
+  ```
+  kubectl apply -f https://raw.githubusercontent.com/ca-gip/kubi/master/deployments/kube-deployment.yml
```
Comment (evrardjp) [inline Command.md:155]: This should be relative to current code, not head of master.

File: Command.md [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+  kubectl apply -f https://raw.githubusercontent.com/ca-gip/kubi/master/deployments/kube-prerequisites.yml
+  kubectl apply -f https://raw.githubusercontent.com/ca-gip/kubi/master/deployments/kube-local-config.yml
+  ```
+  
+
+ Copy the secret from you Kubernetes cluster
+ 
+- Create secret dirs on your local machine
+ ```
+ sudo mkdir -p /var/run/secrets/{certs,ecdsa,kubernetes.io}
```
Comment (evrardjp) [inline Command.md:166]: Why another folder on the host? Please adapt to a temp folder, or a folder inside this git repo, but gitignored?

File: Command.md [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+      - 172.10.0.0/24
+  ingress:
+    # namespaces allowed for ingress rules ( here only nginx )
+    namespaces:
+      - ingress-nginx
+```
+
+** Deploy the example : **
+```bash
+kubectl apply -f https://raw.githubusercontent.com/ca-gip/kubi/master/deployments/kube-example-netpolconf.yml
```
Comment (evrardjp) [inline Command.md:220]: Same thing

File: Command.md [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+current-context: kubi
+contexts:
+- context:
+    cluster:  kubi
+    user: apiserver
+  name: webhook
+```
+
+```bash
+# vim /etc/kubernetes/manifests/kube-apiserver.yaml
```
Comment (evrardjp) [inline Command.md:256]: These two lines seem off. What did you mean ?

File: Readme.md [hunk 19]
```diff
@@ -179,285 +179,13 @@ kubi explain <another_token> for explaining another token
...
-kubectl -n kube-system get secrets kubi -o "jsonpath={.data['tls\.crt']}" | base64 -d > /var/run/secrets/certs/tls.crt
-kubectl -n kube-system get secrets kubi -o "jsonpath={.data['tls\.key']}" | base64 -d > /var/run/secrets/certs/tls.key
-kubectl -n kube-system get secrets kubi-encryption-secret -o "jsonpath={.data['ecdsa-key\.pem']}" | base64 -d > /var/run/secrets/ecdsa/ecdsa-key.pem
-kubectl -n kube-system get secrets kubi-encryption-secret -o "jsonpath={.data['ecdsa-public\.pem']}" | base64 -d > /var/run/secrets/ecdsa/ecdsa-public.pem
-```
-
-
-## Running
-
-At the base of this project execute the go run with the required variable
```
Comment (evrardjp) [inline Readme.md:441]: Great cleanup

File: Readme.md [hunk 20]
```diff
@@ -179,285 +179,13 @@ kubi explain <another_token> for explaining another token
...
+- [Contributing to kubi](https://github.com/amiraaz/kubi/blob/master/CONTRIBUTING.md).
```
Comment (evrardjp) [inline Readme.md:186]: Wrong link

File: deployments/kube-deployment.yml [hunk 21]
```diff
@@ -120,6 +137,9 @@ spec:
         - name: kubi-tls
           secret:
             secretName: kubi
+        - name: kubi-encryption
```
Comment (evrardjp) [inline deployments/kube-deployment.yml:140]: Is this mandatory?

File: deployments/kube-local-config.yml [hunk 22]
```diff
@@ -1,13 +1,13 @@
 apiVersion: v1
 data:
-  LDAP_ADMIN_GROUPBASE: cn=DL_ADMIN_TEAM,OU=GLOBAL,ou=Groups,dc=kubi,dc=ca-gip,dc=github,dc=com
-  LDAP_ADMIN_USERBASE: dc=kubi,dc=ca-gip,dc=github,dc=com
-  LDAP_BINDDN: cn=admin,dc=kubi,dc=ca-gip,dc=github,dc=com
-  LDAP_GROUPBASE: ou=LOCAL,ou=Groups,dc=kubi,dc=ca-gip,dc=github,dc=com
+  LDAP_ADMIN_GROUPBASE: cn=DL_ADMIN_TEAM,OU=GLOBAL,ou=Groups,dc=kubi,dc=cagip,dc=github,dc=com
```
Comment (evrardjp) [inline deployments/kube-local-config.yml:3]: Why the change?

File: deployments/kube-local-config.yml [hunk 23]
```diff
@@ -22,27 +22,27 @@ apiVersion: v1
 kind: ConfigMap
 data:
   kubi.ldif: |
-    dn: ou=People,dc=kubi,dc=ca-gip,dc=github,dc=com
+    dn: ou=People,dc=kubi,dc=c-gip,dc=github,dc=com
```
Comment (evrardjp) [inline deployments/kube-local-config.yml:25]: Incorrect

File: deployments/kube-local-config.yml [hunk 24]
```diff
@@ -22,27 +22,27 @@ apiVersion: v1
...
     objectClass: top
     objectClass: organizationalUnit
     ou: People
 
-    dn: ou=Groups,dc=kubi,dc=ca-gip,dc=github,dc=com
+    dn: ou=Groups,dc=kubi,dc=cagip,dc=github,dc=com
```
Comment (evrardjp) [inline deployments/kube-local-config.yml:30]: Why the change?

**Review by evrardjp — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Can you have a look at my comments please? I am happy to discuss them over chat.

File: Command.md (hunk 13 above)
Comment (amiraaz) [reply Command.md]: should i replace this with a script ?

File: deployments/kube-deployment.yml (hunk 21 above)
Comment (amiraaz) [reply deployments/kube-deployment.yml:140]: yes otherwise it will not be able to find the secret

File: Readme.md (hunk 20 above)
Comment (amiraaz) [reply Readme.md:182]: sorry but i don't understand why it's wrong ?

File: deployments/kube-local-config.yml (hunk 22 above)
Comment (amiraaz) [reply deployments/kube-local-config.yml:3]: the "-" is no longer supported (otherwise the pod will not be started)

**Comment by amiraaz** [comment]
i will make another pull request
